snapcraft/_store.py

```diff
@@ -413,37 +427,100 @@ def push(snap_filename, release_channels=None):
-    else:
-        logger.info('Uploaded {!r}'.format(snap_name))
-    tracker.raise_for_code()
+        logger.info('Revision {!r} of {!r} created'.format(
+            result['revision'], snap_name))\
 
-    if os.environ.get('DELTA_UPLOADS_EXPERIMENTAL'):
-        snap_cache = cache.SnapCache(project_name=snap_name)
-        snap_cache.cache(snap_filename, result['revision'])
-        snap_cache.prune(keep_revision=result['revision'])
+        if os.environ.get('DELTA_UPLOADS_EXPERIMENTAL'):
+            snap_cache.cache(snap_filename, result['revision'])
+            snap_cache.prune(keep_revision=result['revision'])
+    else:
+        logger.info('Pushing {!r}'.format(snap_name))
 
     if release_channels:
         release(snap_name, result['revision'], release_channels)
 
 
+def _push_snap(snap_name, snap_filename):
+    store = storeapi.StoreClient()
+    with _requires_login():
+        tracker = store.upload(snap_name, snap_filename)
+    result = tracker.track()
+    tracker.raise_for_code()
+    return result
+
+
+def _push_delta(snap_name, snap_filename, source_snap):
+    store = storeapi.StoreClient()
+    delta_format = 'xdelta3'
+    logger.info('Found cached source snap {}'.format(source_snap))
+    target_snap = os.path.join(os.getcwd(), snap_filename)
+
+    try:
+        xdelta_generator = deltas.XDelta3Generator(
+            source_path=source_snap, target_path=target_snap)
+        delta_filename = xdelta_generator.make_delta()
+    except (DeltaGenerationError, DeltaGenerationTooBigError):
+        raise StoreDeltaApplicationError
+
+    snap_hashes = {'source_hash': calculate_sha3_384(source_snap),
+                   'target_hash': calculate_sha3_384(target_snap),
+                   'delta_hash': calculate_sha3_384(delta_filename)}
+
+    try:
+        logger.info('Pushing delta {}'.format(delta_filename))
+        with _requires_login():
+            delta_tracker = store.upload(
+                snap_name,
+                delta_filename,
+                delta_format=delta_format,
+                source_hash=snap_hashes['source_hash'],
+                target_hash=snap_hashes['target_hash'],
+                delta_hash=snap_hashes['delta_hash'])
+        result = delta_tracker.track()
+        delta_tracker.raise_for_code()
+    except storeapi.errors.StoreReviewError as e:
+        if e.code == 'processing_upload_delta_error':
+            raise StoreDeltaApplicationError
+        else:
+            raise
+    finally:
+        if os.path.isfile(delta_filename):
+            try:
+                os.remove(delta_filename)
+            except OSError:
+                logger.warning(
+                    'Unable to remove delta {}'.format(delta_filename))
+    return result
+
+
 def _get_text_for_opened_channels(opened_channels):
     if len(opened_channels) == 1:
-        return 'The {!r} channel is now open.'.format(opened_channels[0])
+        return 'The {!r} channel is now open'.format(opened_channels[0])
```

Comment:
While working on this branch iirc, I noticed some inconsistencies in user facing log messages (some have fullstops, others do not). Probably best to discuss and consider for another branch, I'll just revert these.